### ISSUE TITLE
Avoiding SQL Server plan cache pollution due to unparameterized Contains queries generated by EF Core

### ISSUE THREAD

**Issue by ErikEJ** [ErikEJ/erikej.github.io#37]

**Comment by stevendarby** [comment]
Very nice TVF update to my solution here :) https://github.com/dotnet/efcore/issues/13617#issuecomment-716052091

**Comment by ErikEJ** [comment]
Thanks - let me update the post with a credit!

**Comment by ErikEJ** [comment]
Blog post updated, thanks @stevendarby  & @smitpatel

**Comment by garyhalb** [comment]
Hi ErikEJ, finally a seemingly simple way around this issue. Any way this could be implemented to support EF6?

**Comment by joelmandell** [comment]
I have a question about props that is of other type than string? Should I just use overloads or implement with generics then? Or other and better approach?

```
 public IQueryable<string> AsSplit<T>(T[] source)
       => Split(string.Join(",", source), ",").Select(s => s.Value);
```

**Comment by ErikEJ** [comment]
It must be translatable to SQL, there is a sample for long in the blog post

**Comment by ErikEJ** [comment]
@garyhalb is EF6 also not generating correct parameters?

**Comment by garyhalb** [comment]
@ErikEJ Well EF6 has the same plan cache pollution issue that EF Core has. So I was hoping I could use your code sample in my EF6 context, but it doesn't seem to support the FromExpression method in this part:
`[DbFunction(IsBuiltIn = true, Name = "STRING_SPLIT")]
    private IQueryable<StringSplitResult> Split(string source, string separator)
        => FromExpression(() => Split(source, separator));`

**Comment by ErikEJ** [comment]
@garyhalb I will investigate!

**Comment by ErikEJ** [comment]
Got it working, not quite as elegant as EF Core. Blog post coming tomorrow.

**Comment by garyhalb** [comment]
@ErikEJ Fantastic, thanks for your efforts on this!

**Comment by ErikEJ** [comment]
Blog post published: https://erikej.github.io/ef6/sqlserver/2021/12/01/ef6-sqlserver-cache-pollution.html

**Comment by garyhalb** [comment]
@ErikEJ Thanks so much for this, going to get this into testing asap!

**Comment by qmicron** [comment]
Could someone show how to implement STRING_SPLIT for Guid, since there is no way to translate string to Guid.
`public IQueryable<Guid> AsSplit(Guid[] source) => ... .Select(s => ConvertToGuid(s.Value))`

**Comment by ErikEJ** [comment]
@qmicron have you tried:  `new Guid(s.Value)`

**Comment by qmicron** [comment]
No, it could not be translated. The easiest way is just to leave string as output value.
`public IQueryable<string> AsSplit(IEnumerable<Guid> source)
            => Split(string.Join(",", source.Select(x => Convert.ToString(x))), ",").Select(s => s.Value);`

Or more elegant way to make a separate method for GUID converting
`public static Guid? ToGuid(string value) => throw new NotImplementedException();`
Adding registration in OnModelCreating as
`modelBuilder.HasDbFunction(type.GetMethod(nameof(ToGuid))!)
				.HasTranslation(t =>
				{
					return new SqlFunctionExpression(
						functionName: "CONVERT",
						arguments: new[] { new SqlFragmentExpression("UNIQUEIDENTIFIER"), t[0] },
						nullable: true,
						argumentsPropagateNullability: new[] { false, true },
						typeof(Guid?),
						typeMapping: null);
				});`

Then you can write 
`public IQueryable<Guid> AsSplit(IEnumerable<Guid> source)
            => Split(string.Join(",", source.Select(x => Convert.ToString(x))), ",").Select(s => DbFunctions.ToGuid(s.Value).Value);`

It produces following SQL
`SELECT 1 FROM STRING_SPLIT(@__source_1, @__separator_2) AS [s] WHERE CONVERT(UNIQUEIDENTIFIER, [s].[Value]) = [t].[Id])`

Maybe I missed something and there is a better way.

**Comment by ErikEJ** [comment]
@qmicron thanks for sharing a solution

**Comment by jburman** [comment]
If you happen to be working with a case sensitive collation, then you will get an error about column "Value" not being found. The solution is to map StringSplitResult.Value to a lower-case name.  Either with a `[Column("value")]` annotation or with the model builder...

```
modelBuilder.Entity<StringSplitResult>(entity =>
{
    entity.Property(e => e.Value).HasColumnName("value");

    entity.HasNoKey();

}
```

**Comment by ErikEJ** [comment]
@qmicron - fix for the ToGuid method for EF Core 7 (also works with 6) - notice `typeof(Guid)`

```csharp
public static class DbFunctionExtensions
{
    public static Guid? ToGuid(string value) => throw new NotImplementedException();

    public static void ConfigureDbFunctions(this ModelBuilder modelBuilder)
   {
	var type = typeof(DbFunctionExtensions);

        modelBuilder.HasDbFunction(type.GetMethod(nameof(ToGuid))!)
		.HasTranslation(t => new SqlFunctionExpression(
			functionName: "CONVERT",
			arguments: new[] { new SqlFragmentExpression("UNIQUEIDENTIFIER"), t[0] },
			nullable: true,
			argumentsPropagateNullability: new[] { false, true },
			typeof(Guid),
			typeMapping: null))
		.IsBuiltIn();
    }
}
```

**Comment by yv989c** [comment]
Hi 👋. I built QueryableValues to solve this particular problem in SQL Server in a performant way (no more plan cache pollution). https://github.com/yv989c/BlazarTech.QueryableValues

**Comment by netbitshift** [comment]
Im trying to use this example using EF Core 5.x but when writing my query, am getting errors complaining about 'Contains'
eg. IQueryable does not contain a definition for 'Contains'...
